p2p/discovery/store.go

```diff
@@ -0,0 +1,232 @@
+package discovery
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/spacemeshos/go-spacemesh/filesystem"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/p2p/node"
+	"os"
+	"time"
+)
+
+const defaultPeersFileName = "peers.json"
+const saveRoutineInterval = time.Minute * 10
+
+type serializedKnownAddress struct {
+	Addr        string
+	Src         string
+	Attempts    int
+	LastSeen    int64
+	LastAttempt int64
+	LastSuccess int64
+	// no refcount or tried, that is available from context.
+}
+
+type serializedAddrManager struct {
+	Key          [32]byte
+	Addresses    []*serializedKnownAddress
+	NewBuckets   [newBucketCount][]string // NodeInfo represented as string
+	TriedBuckets [triedBucketCount][]string
+}
+
+// savePeers saves all the known addresses to a file so they can be read back
+// in at next run.
+func (a *addrBook) savePeers(path string) {
+	a.mtx.Lock()
+	defer a.mtx.Unlock()
+
+	// First we make a serialisable datastructure so we can encode it to
+	// json.
+	sam := new(serializedAddrManager)
+	copy(sam.Key[:], a.key[:])
+
+	sam.Addresses = make([]*serializedKnownAddress, len(a.addrIndex))
+	i := 0
+	for _, v := range a.addrIndex {
+		ska := new(serializedKnownAddress)
+		ska.Addr = v.na.String()
+		ska.Src = v.srcAddr.String()
+		ska.LastSeen = v.lastSeen.Unix()
+		ska.Attempts = v.attempts
+		ska.LastAttempt = v.lastattempt.Unix()
+		ska.LastSuccess = v.lastsuccess.Unix()
+		// Tried and refs are implicit in the rest of the structure
+		// and will be worked out from context on unserialisation.
+		sam.Addresses[i] = ska
+		i++
+	}
+	for i := range a.addrNew {
+		sam.NewBuckets[i] = make([]string, len(a.addrNew[i]))
+		j := 0
+		for k := range a.addrNew[i] {
+			sam.NewBuckets[i][j] = k
+			j++
+		}
+	}
+	for i := range a.addrTried {
+		sam.TriedBuckets[i] = make([]string, len(a.addrTried[i]))
+		j := 0
+		for k := range a.addrTried[i] {
+			sam.TriedBuckets[i][j] = k
+			j++
+		}
+	}
+
+	w, err := os.Create(path)
+	if err != nil {
+		log.Error("Error opening file %s: %v", path, err)
+		return
+	}
+	enc := json.NewEncoder(w)
+	defer w.Close()
+	if err := enc.Encode(&sam); err != nil {
+		log.Error("Failed to encode file %s: %v", path, err)
+		return
+	}
+}
+
+// loadPeers loads the known address from the saved file.  If empty, missing, or
+// malformed file, just don't load anything and start fresh
+func (a *addrBook) loadPeers(filePath string) {
+	a.mtx.Lock()
```

Comment:
consider moving locks to deserializePeers

Reply:
we don't lock since loading might fail and then we'll `reset`. added a comment about it